### PR TITLE
fix(workspace): race condition when backing up configuration 

### PR DESCRIPTION
**Pull request by kevinslin**
fix(workspace): race condition when backing up configuration 

When ensuring that a backup folder exists, actually await the return result

***
# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

## Instrumentation

NA

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [~] Common cases tested
- [~] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended

NA

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [x] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

***

Doc updates: https://github.com/dendronhq/dendron-docs/commit/84ae197d0729315979d54920f0d9b52a3cc13d7e

### CODE REVIEW

File: packages/engine-server/src/backup/service.ts [hunk 1]
```diff
@@ -68,8 +69,8 @@ export class BackupService implements Disposable, IBackupService {
       root: this.wsRoot,
     });
     await fs.ensureDir(this.backupRoot);
-    Object.keys(BackupKeyEnum).forEach(async (key) => {
-      await fs.ensureDir(path.join(this.backupRoot, key));
+    await asyncLoopOneAtATime<string>(Object.values(BackupKeyEnum), (key) => {
+      return fs.ensureDir(path.join(this.backupRoot, key));
```
Comment (jonathanyeung) [inline packages/engine-server/src/backup/service.ts:73]: In this case, do we actually need the logic in the loop to be sequential though?

i.e. versus
```typescript
    await Promise.all(
      Object.keys(BackupKeyEnum).map((key) =>
        fs.ensureDir(path.join(this.backupRoot, key))
      )
    );
```

Comment (kevinslin) [reply packages/engine-server/src/backup/service.ts:73]: that's fair. updated

**Comment by hikchoi** [comment]
Thanks for the catch. The ensureDir calls were originally sync, but I changed it last minute to async and forgot to stick a promise.all on them 😓

**Review by jonathanyeung — APPROVED** [review]

File: packages/engine-server/src/backup/service.ts [hunk 2]
```diff
@@ -68,8 +69,8 @@ export class BackupService implements Disposable, IBackupService {
       root: this.wsRoot,
     });
     await fs.ensureDir(this.backupRoot);
-    Object.keys(BackupKeyEnum).forEach(async (key) => {
-      await fs.ensureDir(path.join(this.backupRoot, key));
+    await asyncLoop<string>(Object.values(BackupKeyEnum), (key) => {
```
Comment (SeriousBug) [inline packages/engine-server/src/backup/service.ts:72]: Could we add a docstring to `BackupKeyEnum` to say whether it's the key or the value that's supposed to be the path? The switch from the key to value confused me when reviewing at least

**Review by SeriousBug — APPROVED** [review]
Looks good, one comment for documentation

**Review by hikchoi — APPROVED** [review]

**Pull request merged**